net: change CNetAddr::ip to have flexible size

vasild opened the pull request:
(chopped off from #19031 to ease review)

Before this change `CNetAddr::ip` was a fixed-size array of 16 bytes,
not being able to store larger addresses (e.g. TORv3) and encoded
smaller ones as 16-byte IPv6 addresses.

Change its type to `prevector`, so that it can hold larger addresses and
do not disguise non-IPv6 addresses as IPv6. So the IPv4 address
`1.2.3.4` is now encoded as `01020304` instead of
`00000000000000000000FFFF01020304`.

Rename `CNetAddr::ip` to `CNetAddr::m_addr` because it is not an "IP" or
"IP address" (TOR addresses are not IP addresses).

In order to preserve backward compatibility with serialization (where
e.g. `1.2.3.4` is serialized as `00000000000000000000FFFF01020304`)
introduce `CNetAddr` dedicated legacy serialize/unserialize methods.

Adjust `CSubNet` accordingly. Still use `CSubNet::netmask[]` of fixed 16
bytes, but use the first 4 for IPv4 (not the last 4). Do not accept
invalid netmasks that have 0-bits followed by 1-bits and only allow
subnetting for IPv4 and IPv6.

Co-authored-by: Carl Dong <contact@carldong.me>

jonatack commented:
Concept ACK

vasild commented:
[Filtered code coverage report](https://people.freebsd.org/~vd/pr19628_daa1d8b11_coverage/) (files not modified by this PR are omitted and not modified lines in files that are otherwise modified are dimmed).

[List of modified and not covered lines](https://people.freebsd.org/~vd/pr19628_daa1d8b11_coverage/modified_and_not_covered.html).

practicalswift commented on src/netaddress.h:
```diff
@@ -49,18 +51,50 @@ enum Network
+         * Raw representation of the network address.
+         * In network byte order (big endian) for IPv4 and IPv6.
+         */
+        prevector<ADDR_IPv6_SIZE, uint8_t> m_addr{ADDR_IPv6_SIZE, 0x0};
```
Are you sure the use of `prevector` is warranted here? Measurements would be nice.

Generally I think we should try stick with to the standard containers unless we have clear quantitative evidence showing that they are unsuitable for the scenario being considered. Reviewers, static analysers and compilers tend to be better at analysing use of the standard containers.

FWIW: 

```
$ git grep -l prevector -- "*.cpp" "*.h" ":(exclude)src/bench/" ":(exclude)src/test/"
src/hash.h
src/memusage.h
src/prevector.h
src/script/script.h
src/serialize.h
```

sipa replied:
I suggested this, so let me back it up.

Currently an address in CAddrMan consumes 256 bytes (cost in mapInfo and mapAddr together, plus allocation overhead).

This PR as-is raises that to 288 bytes, for addresses that take up to 16 bytes (which is all of them, for now).

Using an `std::vector<uint8_t>` instead would change that to 352 bytes per address, or an extra 64 on top. It would also increase memory fragmentation and (slightly) increase CPU usage because of two malloc's per address.

FTR, my node has 69691 addresses in CAddrMan.

practicalswift replied:
@sipa Thanks for quantifying! That answers my question :)

vasild replied:
Thanks for the question and the elaborate explanation! Indeed smaller size and avoiding an extra heap allocation is better. FWIW I moved `m_addr` definition before `m_net` to save 8 bytes from the size of `CNetAddr` (32 vs 40 bytes on x64).

sipa commented on src/netaddress.cpp:
```diff
@@ -85,151 +70,161 @@ bool CNetAddr::SetInternal(const std::string &name)
+    if (!IsIPv4() && !IsIPv6()) {
+        return false;
+    }
+    for (uint8_t b : m_addr) {
```
Slightly more modern:

```c++
return std::all_of(m_addr.begin(), m_addr.end(), [](uint8_t b) { return b == 0; });
```

vasild replied:
Changed.

sipa commented on src/netaddress.cpp:
```diff
@@ -287,17 +283,12 @@ bool CNetAddr::IsValid() const
-        if (memcmp(ip+12, &ipNone, 4) == 0)
-            return false;
+    if (IsIPv4()) {
+        for (uint32_t a : {(uint32_t)INADDR_ANY, (uint32_t)INADDR_NONE}) {
```
Not something you introduced, but this only works because those addresses have the same byte representation in every byte order.

vasild replied:
Right! Even though the values of those constants is never going to change (I hope!), does the following look better?

```cpp
for (uint32_t a : {htonl(INADDR_ANY), htonl(INADDR_NONE)}) {
```

laanwj replied:
That seems the best option, I think. It's too bad that there's no way to get these constants as byte spans that can be directly compared against m_addr so it wouldn't need the cast and `memcmp` at all.

vasild replied:
Changed to use `htonl()`.

sipa commented on src/netaddress.cpp:
```diff
@@ -369,12 +360,12 @@ std::string CNetAddr::ToString() const
 bool operator<(const CNetAddr& a, const CNetAddr& b)
 {
-    return a.m_net < b.m_net || (a.m_net == b.m_net && memcmp(a.ip, b.ip, 16) < 0);
+    return a.m_net < b.m_net || (a.m_net == b.m_net && a.m_addr < b.m_addr);
```
More modern:

```c++
return std::tie(a.m_net, a.m_addr) < std::tie(b.m_net, b.m_addr);
```

vasild replied:
Modernized :bulb:

sipa commented on src/netaddress.h:
```diff
@@ -49,18 +51,50 @@ enum Network
 };
 
+/// If an IPv6 address begins with this, then we treat the rest of it as IPv4 address.
+static constexpr uint8_t IPv4_IN_IPv6_PREFIX[12] = {
```
Nit: uppercase v in global constants (here and further).

vasild replied:
Changed to uppercase.

vasild commented:
Rebased to resolve a conflict.

vasild commented:
Applied review suggestions.

laanwj commented on src/test/fuzz/asmap.cpp:
```diff
@@ -43,7 +43,17 @@ void test_one_input(const std::vector<uint8_t>& buffer)
-    net_addr.SetRaw(ipv6 ? NET_IPV6 : NET_IPV4, buffer.data() + 1 + asmap_size);
+    if (ipv6) {
+        assert(addr_size == ADDR_IPV6_SIZE);
+        net_addr.SetLegacyIPv6(*reinterpret_cast<const uint8_t (*)[ADDR_IPV6_SIZE]>(addr_data));
```
This is pretty scary cast expression. Wouldn't a cast to just `const uint8_t*` work just as well?

vasild replied:
Passing `const uint8_t*` will not work as long as the argument of the method is a reference to an array: `CNetAddr::SetLegacyIPv6(const uint8_t (&ipv6)[ADDR_IPV6_SIZE])` which is best of type safety but unfortunately two of the callers don't have an array to pass.

I think either one of the following two is better:

<details>
<summary>change argument to void*</summary>

```diff
diff --git i/src/netaddress.cpp w/src/netaddress.cpp
index b421e1b11..bb4d2b50b 100644
--- i/src/netaddress.cpp
+++ w/src/netaddress.cpp
@@ -26,32 +26,33 @@ CNetAddr::CNetAddr() {}
 void CNetAddr::SetIP(const CNetAddr& ipIn)
 {
     m_net = ipIn.m_net;
     m_addr = ipIn.m_addr;
 }
 
-void CNetAddr::SetLegacyIPv6(const uint8_t (&ipv6)[ADDR_IPV6_SIZE])
+void CNetAddr::SetLegacyIPv6(const void* ipv6)
 {
-    const uint8_t* ipv6_end = ipv6 + ADDR_IPV6_SIZE;
+    const uint8_t* ipv6_begin = static_cast<const uint8_t*>(ipv6);
+    const uint8_t* ipv6_end = ipv6_begin + ADDR_IPV6_SIZE;
 
-    if (memcmp(ipv6, IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
+    if (memcmp(ipv6_begin, IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
         // IPv4-in-IPv6
         m_net = NET_IPV4;
-        m_addr.assign(ipv6 + sizeof(IPV4_IN_IPV6_PREFIX), ipv6_end);
-    } else if (memcmp(ipv6, TORV2_IN_IPV6_PREFIX, sizeof(TORV2_IN_IPV6_PREFIX)) == 0) {
+        m_addr.assign(ipv6_begin + sizeof(IPV4_IN_IPV6_PREFIX), ipv6_end);
+    } else if (memcmp(ipv6_begin, TORV2_IN_IPV6_PREFIX, sizeof(TORV2_IN_IPV6_PREFIX)) == 0) {
         // TORv2-in-IPv6
         m_net = NET_ONION;
-        m_addr.assign(ipv6 + sizeof(TORV2_IN_IPV6_PREFIX), ipv6_end);
-    } else if (memcmp(ipv6, INTERNAL_IN_IPV6_PREFIX, sizeof(INTERNAL_IN_IPV6_PREFIX)) == 0) {
+        m_addr.assign(ipv6_begin + sizeof(TORV2_IN_IPV6_PREFIX), ipv6_end);
+    } else if (memcmp(ipv6_begin, INTERNAL_IN_IPV6_PREFIX, sizeof(INTERNAL_IN_IPV6_PREFIX)) == 0) {
         // Internal-in-IPv6
         m_net = NET_INTERNAL;
-        m_addr.assign(ipv6 + sizeof(INTERNAL_IN_IPV6_PREFIX), ipv6_end);
+        m_addr.assign(ipv6_begin + sizeof(INTERNAL_IN_IPV6_PREFIX), ipv6_end);
     } else {
         // IPv6
         m_net = NET_IPV6;
-        m_addr.assign(ipv6, ipv6_end);
+        m_addr.assign(ipv6_begin, ipv6_end);
     }
 }
 
 /**
  * Try to make this a dummy address that maps the specified name into IPv6 like
  * so: (0xFD + %sha256("bitcoin")[0:5]) + %sha256(name)[0:10]. Such dummy
@@ -105,13 +106,13 @@ CNetAddr::CNetAddr(const struct in_addr& ipv4Addr)
     m_addr.assign(ptr, ptr + ADDR_IPV4_SIZE);
 }
 
 CNetAddr::CNetAddr(const struct in6_addr& ipv6Addr, const uint32_t scope)
 {
     assert(sizeof(ipv6Addr) == ADDR_IPV6_SIZE);
-    SetLegacyIPv6(*reinterpret_cast<const uint8_t (*)[ADDR_IPV6_SIZE]>(&ipv6Addr));
+    SetLegacyIPv6(&ipv6Addr);
     scopeId = scope;
 }
 
 bool CNetAddr::IsBindAny() const
 {
     if (!IsIPv4() && !IsIPv6()) {
diff --git i/src/netaddress.h w/src/netaddress.h
index 8924ed0b4..ca048fbf5 100644
--- i/src/netaddress.h
+++ w/src/netaddress.h
@@ -105,13 +105,13 @@ class CNetAddr
         /**
          * Set from a legacy IPv6 address.
          * Legacy IPv6 address may be a normal IPv6 address, or another address
          * (e.g. IPv4) disguised as IPv6. This encoding is used in the legacy
          * `addr` encoding.
          */
-        void SetLegacyIPv6(const uint8_t (&ipv6)[ADDR_IPV6_SIZE]);
+        void SetLegacyIPv6(const void* ipv6);
 
         bool SetInternal(const std::string& name);
 
         bool SetSpecial(const std::string &strName); // for Tor addresses
         bool IsBindAny() const; // INADDR_ANY equivalent
         bool IsIPv4() const;    // IPv4 mapped address (::FFFF:0:0/96, 0.0.0.0/0)
diff --git i/src/test/fuzz/asmap.cpp w/src/test/fuzz/asmap.cpp
index 84902ee2d..98ac82453 100644
--- i/src/test/fuzz/asmap.cpp
+++ w/src/test/fuzz/asmap.cpp
@@ -45,13 +45,13 @@ void test_one_input(const std::vector<uint8_t>& buffer)
     if (!SanityCheckASMap(asmap)) return;
 
     const uint8_t* addr_data = buffer.data() + 1 + asmap_size;
     CNetAddr net_addr;
     if (ipv6) {
         assert(addr_size == ADDR_IPV6_SIZE);
-        net_addr.SetLegacyIPv6(*reinterpret_cast<const uint8_t (*)[ADDR_IPV6_SIZE]>(addr_data));
+        net_addr.SetLegacyIPv6(addr_data);
     } else {
         assert(addr_size == ADDR_IPV4_SIZE);
         in_addr ipv4;
         memcpy(&ipv4, addr_data, addr_size);
         net_addr.SetIP(CNetAddr{ipv4});
     }
```
</details>

<details>
<summary>change argument to Span</summary>

```diff
diff --git i/src/netaddress.cpp w/src/netaddress.cpp
index b421e1b11..121b92650 100644
--- i/src/netaddress.cpp
+++ w/src/netaddress.cpp
@@ -26,32 +26,32 @@ CNetAddr::CNetAddr() {}
 void CNetAddr::SetIP(const CNetAddr& ipIn)
 {
     m_net = ipIn.m_net;
     m_addr = ipIn.m_addr;
 }
 
-void CNetAddr::SetLegacyIPv6(const uint8_t (&ipv6)[ADDR_IPV6_SIZE])
+void CNetAddr::SetLegacyIPv6(Span<const uint8_t> ipv6)
 {
-    const uint8_t* ipv6_end = ipv6 + ADDR_IPV6_SIZE;
+    assert(ipv6.size() == ADDR_IPV6_SIZE);
 
-    if (memcmp(ipv6, IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
+    if (memcmp(ipv6.begin(), IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
         // IPv4-in-IPv6
         m_net = NET_IPV4;
-        m_addr.assign(ipv6 + sizeof(IPV4_IN_IPV6_PREFIX), ipv6_end);
-    } else if (memcmp(ipv6, TORV2_IN_IPV6_PREFIX, sizeof(TORV2_IN_IPV6_PREFIX)) == 0) {
+        m_addr.assign(ipv6.begin() + sizeof(IPV4_IN_IPV6_PREFIX), ipv6.end());
+    } else if (memcmp(ipv6.begin(), TORV2_IN_IPV6_PREFIX, sizeof(TORV2_IN_IPV6_PREFIX)) == 0) {
         // TORv2-in-IPv6
         m_net = NET_ONION;
-        m_addr.assign(ipv6 + sizeof(TORV2_IN_IPV6_PREFIX), ipv6_end);
-    } else if (memcmp(ipv6, INTERNAL_IN_IPV6_PREFIX, sizeof(INTERNAL_IN_IPV6_PREFIX)) == 0) {
+        m_addr.assign(ipv6.begin() + sizeof(TORV2_IN_IPV6_PREFIX), ipv6.end());
+    } else if (memcmp(ipv6.begin(), INTERNAL_IN_IPV6_PREFIX, sizeof(INTERNAL_IN_IPV6_PREFIX)) == 0) {
         // Internal-in-IPv6
         m_net = NET_INTERNAL;
-        m_addr.assign(ipv6 + sizeof(INTERNAL_IN_IPV6_PREFIX), ipv6_end);
+        m_addr.assign(ipv6.begin() + sizeof(INTERNAL_IN_IPV6_PREFIX), ipv6.end());
     } else {
         // IPv6
         m_net = NET_IPV6;
-        m_addr.assign(ipv6, ipv6_end);
+        m_addr.assign(ipv6.begin(), ipv6.end());
     }
 }
 
 /**
  * Try to make this a dummy address that maps the specified name into IPv6 like
  * so: (0xFD + %sha256("bitcoin")[0:5]) + %sha256(name)[0:10]. Such dummy
@@ -104,14 +104,13 @@ CNetAddr::CNetAddr(const struct in_addr& ipv4Addr)
     const uint8_t* ptr = reinterpret_cast<const uint8_t*>(&ipv4Addr);
     m_addr.assign(ptr, ptr + ADDR_IPV4_SIZE);
 }
 
 CNetAddr::CNetAddr(const struct in6_addr& ipv6Addr, const uint32_t scope)
 {
-    assert(sizeof(ipv6Addr) == ADDR_IPV6_SIZE);
-    SetLegacyIPv6(*reinterpret_cast<const uint8_t (*)[ADDR_IPV6_SIZE]>(&ipv6Addr));
+    SetLegacyIPv6(Span<const uint8_t>(reinterpret_cast<const uint8_t*>(&ipv6Addr), sizeof(ipv6Addr)));
     scopeId = scope;
 }
 
 bool CNetAddr::IsBindAny() const
 {
     if (!IsIPv4() && !IsIPv6()) {
diff --git i/src/netaddress.h w/src/netaddress.h
index 8924ed0b4..a76b5e9ce 100644
--- i/src/netaddress.h
+++ w/src/netaddress.h
@@ -105,13 +105,13 @@ class CNetAddr
         /**
          * Set from a legacy IPv6 address.
          * Legacy IPv6 address may be a normal IPv6 address, or another address
          * (e.g. IPv4) disguised as IPv6. This encoding is used in the legacy
          * `addr` encoding.
          */
-        void SetLegacyIPv6(const uint8_t (&ipv6)[ADDR_IPV6_SIZE]);
+        void SetLegacyIPv6(Span<const uint8_t> ipv6);
 
         bool SetInternal(const std::string& name);
 
         bool SetSpecial(const std::string &strName); // for Tor addresses
         bool IsBindAny() const; // INADDR_ANY equivalent
         bool IsIPv4() const;    // IPv4 mapped address (::FFFF:0:0/96, 0.0.0.0/0)
diff --git i/src/test/fuzz/asmap.cpp w/src/test/fuzz/asmap.cpp
index 84902ee2d..e3aefa18a 100644
--- i/src/test/fuzz/asmap.cpp
+++ w/src/test/fuzz/asmap.cpp
@@ -45,13 +45,13 @@ void test_one_input(const std::vector<uint8_t>& buffer)
     if (!SanityCheckASMap(asmap)) return;
 
     const uint8_t* addr_data = buffer.data() + 1 + asmap_size;
     CNetAddr net_addr;
     if (ipv6) {
         assert(addr_size == ADDR_IPV6_SIZE);
-        net_addr.SetLegacyIPv6(*reinterpret_cast<const uint8_t (*)[ADDR_IPV6_SIZE]>(addr_data));
+        net_addr.SetLegacyIPv6(Span<const uint8_t>(addr_data, addr_size));
     } else {
         assert(addr_size == ADDR_IPV4_SIZE);
         in_addr ipv4;
         memcpy(&ipv4, addr_data, addr_size);
         net_addr.SetIP(CNetAddr{ipv4});
     }
```
</details>

What do you think?

sipa replied:
Span span span...

![span](https://user-images.githubusercontent.com/548488/89688260-a0b3b200-d8b6-11ea-9f9c-6b8c7f87b385.png)

vasild replied:
Done

laanwj commented on src/netaddress.cpp:
```diff
@@ -341,9 +329,9 @@ enum Network CNetAddr::GetNetwork() const
+        return EncodeBase32(m_addr.data(), m_addr.size()) + ".onion";
     if (IsInternal())
-        return EncodeBase32(ip + sizeof(g_internal_prefix), sizeof(ip) - sizeof(g_internal_prefix)) + ".internal";
+        return EncodeBase32(m_addr.data(), m_addr.size()) + ".internal";
```
Heh we need to make this one take a span too (not in this PR ofcourse).

vasild replied:
Done in https://github.com/bitcoin/bitcoin/pull/19687

laanwj commented on src/netaddress.cpp:
```diff
@@ -822,68 +851,27 @@ bool CSubNet::Match(const CNetAddr &addr) const
     }
 
-    return network.ToString() + "/" + strNetmask;
+    return network.ToString() + "/" + strprintf("%hhu", cidr);
```
You do not need the "hh" here, or any type specifiers ever, `strprintf` is type safe.
Also maybe include the `/` to avoid an extra string concatenation
```
network.ToString() + strprintf("/%u", cidr);
```

vasild replied:
Changed.

vasild commented:
Fixed Windows build and changed `SetLegacyIPv6()` to take a ~~Spam~~Span argument.

vasild commented:
`s/IPv4_IN_IPv6_PREFIX/IPV4_IN_IPV6_PREFIX/` also in comment, was missed during the rename.

vasild commented:
Rebased just to restart Travis.

jonatack reviewed (commented):
Glad to see progress on BIP155/addrv2 with the imminent Tor v2 deprecation in a month with planned obsolescence on July 15, 2021.

ACK bfe7d1b56500b08bb a couple hours of code review -- tried to distinguish the many straightforward-looking and good mechanical changes from the critical ones as they are all smooshed together, however I will continue looking at this code and it's good to have it merged not late in the release cycle--, clean debug builds on gcc Debian 10.1.0, running a node since a day now with the change.

A few comments below that can be ignored or left for a follow-up unless you decide to retouch or take some of them now.

vasild commented on src/netaddress.cpp:
```diff
@@ -507,16 +503,14 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
     vchRet.push_back(net_class);
-    int nStartByte = 0;
-    int nBits = 16;
+    int nBits;
```
Default assigning in the case where we are supposed to always assign the variable later will suppress valgrind and compiler warnings if there is a bug in the code and it does not actually initialize the variable in some cases.

jonatack replied:
Some history and reading on uninitialized reads in Bitcoin Core: https://bitcoincore.reviews/17639

See also this comment and thread https://github.com/bitcoin/bitcoin/pull/17627#issuecomment-559421439

vasild replied:
Initialized to `0`. Note that the Cpp Core Guidelines contain an [example](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#example-186) that resembles our case and admits that "This cannot trivially be rewritten to initialize i and j with initializers".

We have:

```cpp
int nBits;
... code that is supposed to set nBits ... // [1]
read nBits // [2]
```

So if there is a bug in [1] and `nBits` is not set (or if new code is added that reads `nBits` before [1]), then we would get undefined behavior in [2], but we would also get compiler, msan and valgrind warnings about it, so not likely to remain unnoticed.

OTOH if we initialize `nBits` with some `dummy_value` and the same bugs happen, then we would not get undefined behavior, but we would get unexpected results due to [2] reading `dummy_value`. And the downside is that none of the compiler, msan or valgrind will warn about the bug.

jonatack replied:
Yes. Attempted to summarize the trade-offs you point out in the review club notes I linked to (LMK if you see any errors, I wanted it to be a resource).

MarcoFalke replied:
Maybe it is time to add an init macro that default initializes a variable, but disables the initialization when `--enable-debug` or so?

vasild commented on src/netaddress.h:
```diff
@@ -152,14 +180,91 @@ class CNetAddr
+                break;
+            case NET_ONION:
+                prefix_size = sizeof(TORV2_IN_IPV6_PREFIX);
+                assert(m_addr.size() + prefix_size == sizeof(arr));
```
Done.

vasild commented on src/netaddress.h:
```diff
@@ -152,14 +180,91 @@ class CNetAddr
+            case NET_MAX:
+                memset(arr, 0x0, sizeof(arr));
+                break;
+            }
```
Done.

jonatack replied:
Add the assert as well as the comment?

vasild replied:
Hmm, I almost added the `assert(false);` but then realized that the code flow just continues after the `switch` (it contains `break;`, not `return;` as the example in the developer-notes).

Apply this?

<details>
<summary>diff</summary>

```diff
diff --git i/src/netaddress.h w/src/netaddress.h
index b75eff1b9..8cd2401e2 100644
--- i/src/netaddress.h
+++ w/src/netaddress.h
@@ -200,36 +200,38 @@ class CNetAddr
             size_t prefix_size;
 
             switch (m_net) {
             case NET_IPV6:
                 assert(m_addr.size() == sizeof(arr));
                 memcpy(arr, m_addr.data(), m_addr.size());
-                break;
+                return;
             case NET_IPV4:
                 prefix_size = sizeof(IPV4_IN_IPV6_PREFIX);
                 assert(prefix_size + m_addr.size() == sizeof(arr));
                 memcpy(arr, IPV4_IN_IPV6_PREFIX, prefix_size);
                 memcpy(arr + prefix_size, m_addr.data(), m_addr.size());
-                break;
+                return;
             case NET_ONION:
                 prefix_size = sizeof(TORV2_IN_IPV6_PREFIX);
                 assert(prefix_size + m_addr.size() == sizeof(arr));
                 memcpy(arr, TORV2_IN_IPV6_PREFIX, prefix_size);
                 memcpy(arr + prefix_size, m_addr.data(), m_addr.size());
-                break;
+                return;
             case NET_INTERNAL:
                 prefix_size = sizeof(INTERNAL_IN_IPV6_PREFIX);
                 assert(prefix_size + m_addr.size() == sizeof(arr));
                 memcpy(arr, INTERNAL_IN_IPV6_PREFIX, prefix_size);
                 memcpy(arr + prefix_size, m_addr.data(), m_addr.size());
-                break;
+                return;
             case NET_UNROUTABLE:
             case NET_MAX:
                 memset(arr, 0x0, sizeof(arr));
-                break;
+                return;
             } // no default case, so the compiler can warn about missing cases
+
+            assert(false);
         }
 
         /**
          * Serialize in pre-ADDRv2/BIP155 format to a stream.
          * Some addresses (e.g. TORv3) cannot be serialized in pre-BIP155 format.
          */
```
</details>

jonatack replied:
Returning LGTM, nothing else is being done in the function.

vasild replied:
Changed `break;` to `return;` and added `assert(false);` after the `switch`.

vasild commented on src/netaddress.cpp:
```diff
@@ -538,22 +531,33 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
-        nStartByte++;
+        assert(i < m_addr.size());
+        vchRet.push_back(m_addr[i]);
+        i++;
```
Done.

vasild commented:
Addressed review suggestions.

vasild commented:
Addressed review suggestion.

vasild commented:
Addressed review suggestion.

jonatack commented:
re-ACK 8b7f2c1 per `git diff bfe7d1b 8b7f2c1`

Ran bitcoind at bfe7d1b with both onion and clearnet ipv4/ipv6 conns open for a couple of days, debug log and `-netinfo` connections dashboard have been showing nominal operation. Restarting bitcoind now at 8b7f2c1.

prusnak commented:
What about TORv3 addresses?

prusnak commented:
I assume TORv3 addresses will be addressed in the future PRs, right?

laanwj commented:
> I assume TORv3 addresses will be addressed in the future PRs, right?

Yes. See BIP155 and  #19031. This is just preparatory work required for that,

elichai commented on src/netaddress.cpp:
```diff
@@ -3,63 +3,57 @@
-    if (memcmp(ipv6, pchIPv4, sizeof(pchIPv4)) == 0) {
+    assert(ipv6.size() == ADDR_IPV6_SIZE);
+
+    if (memcmp(ipv6.begin(), IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
```
nit, you could reformat this to more "modern" C++ (though I'm not sure if it is less or more readable)
```c++
if(std::equal(std::cbegin(IPV4_IN_IPV6_PREFIX), std::cend(IPV4_IN_IPV6_PREFIX), ipv6.cbegin()))
```

sipa replied:
Agree with @elichai's suggestion, `std::equal` is nicer.

Given how many of these tests are being done (including the IsRFC... stuff below), perhaps you want to go even further (this is absolutely not a blocker for this PR, however):

```c++
template<typename T, int LEN>
bool HasPrefix(const T1& obj, const uint8_t (&prefix)[LEN])
{
    return obj.size() >= LEN && std::equal(std::cbegin(prefix), std::cend(prefix), std::cbegin(obj));
}

...

if (HasPrefix(ipv6, IPV4_IN_IPV6_PREFIX)) ...
```

If you turn the prefixes into `std::array` it could be
```c++
template<typename T1, typename T2>
bool HasPrefix(const T1& obj, const T2& prefix)
{
    return obj.size() >= prefix.size() && std::equal(std::cbegin(prefix), std::cend(prefix), std::cbegin(obj));
}
```

instead.

elichai replied:
In C++17 `HasPrefix` will be possible even with C arrays :) (by replacing `.size()` with `std::size(..)`)

vasild replied:
Done - [diff together with other suggestions](https://github.com/bitcoin/bitcoin/compare/8b7f2c13b67ae84a59f29a7cffe243055a2b93d9..86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3).

kallewoof commented on src/netaddress.cpp:
```diff
@@ -85,151 +79,154 @@ bool CNetAddr::SetInternal(const std::string &name)
+    if (!IsIPv4() && !IsIPv6()) {
+        return false;
+    }
+    return std::all_of(m_addr.begin(), m_addr.end(), [](uint8_t b) { return b == 0; });
```
Shouldn't this be `!std::all_of(...)`?

kallewoof commented on src/netaddress.cpp:
```diff
@@ -85,151 +79,154 @@ bool CNetAddr::SetInternal(const std::string &name)
-        if (GetByte(i)) return false;
+    if (m_net == NET_IPV4) {
+        assert(m_addr.size() == ADDR_IPV4_SIZE);
+        return true;
```
Placing this assertion here seems weird to me. Shouldn't this check be done when(ever) `m_net` is modified? Same with `IsIPv6()` below.

vasild replied:
Done - [diff together with other suggestions](https://github.com/bitcoin/bitcoin/compare/8b7f2c13b67ae84a59f29a7cffe243055a2b93d9..86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3).

kallewoof commented on src/netaddress.cpp:
```diff
@@ -287,17 +285,12 @@ bool CNetAddr::IsValid() const
+            if (memcmp(m_addr.data(), &a, sizeof(a)) == 0) {
+                return false;
+            }
+        }
```
The opposite seems more straightforward:
```C++
if (IsIPv4()) {
    uint32_t a;
    memcpy(a, m_addr.data(), sizeof(a));
    if (a == htonl(INADDR_ANY) || a == htonl(INADDR_NONE)) return false;
}
```
@sipa might yell at me but it might also work to do `uint32_t a = *(uint32_t*)m_addr.data();` to skip the `memcpy` line.

vasild replied:
Those constants are defined like this in the system headers:

```cpp
typedef uint32_t in_addr_t;
...
#define	INADDR_ANY		((in_addr_t) 0x00000000)
#define	INADDR_NONE		((in_addr_t) 0xffffffff)
...
#define INADDR_ALLHOSTS_GROUP	((in_addr_t) 0xe0000001) /* 224.0.0.1 */
```

Of course any discussion about the byte order of `INADDR_ANY` is meaningless, but nevertheless - we can assume that all `INADDR_*` are in the same byte order. From the above we see that 224.0.0.1 is represented as the integer `3758096385` or `0xe0000001` which on LE systems is stored in memory as `01 00 00 e0` and on BE systems as `e0 00 00 01`.

This means that whenever we want to compare those constants to something we know is in BE byte order (like `m_addr`) we have to apply `htonl()` to them. This is how it is done in the [L](http://tomoyo.osdn.jp/cgi-bin/lxr/source/net/bridge/br_multicast.c?a=sparc64#L263)inu[x](http://tomoyo.osdn.jp/cgi-bin/lxr/source/net/ipv4/igmp.c?a=sparc64#L1542) and [FreeBSD](https://github.com/freebsd/freebsd/blob/300d08e9968a2e38e163d3c3479524a587fe8577/sys/netinet/in.h#L658) system headers.

So, leaving this as is at this iteration.

sipa replied:
You're right.

I was confused by the fact that `struct in_addr`'s `s_addr` field stores IP addresses in BE order, but the INADDR_* constants are in native order (and you are supposed to use `htonl` & friends to convert between them; see the section on `sin_addr` in https://man7.org/linux/man-pages/man7/ip.7.html).

Arguably you should be using `ntohl` instead of `htonl` (as these constants are in native order, and we want them in BE order) - but on every realistic platform these two are equivalent.

If you want to avoid the winsock/inet header dependency you can also use WriteBE32 to convert them to bytes.

elichai replied:
as @sipa said 
```C++
uint32_t a = ReadBE32(m_data.data());
if (a == INADDR_ANY || a == INADDR_ALL) ...
```
is so much more readable :)

sipa replied:
Oops, yes, I guess this confused @kallewoof too (on IRC). Specifially, this is UB:

```c++
uint32_t a = *(uint32_t*)m_addr.data();
```

as it's accessing a byte array through an incompatible type.

kallewoof commented on src/netaddress.cpp:
```diff
@@ -538,22 +529,33 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
-        nStartByte++;
+        assert(i < m_addr.size());
+        vchRet.push_back(m_addr[i]);
+        ++i;
```
It looks like you can do this as a copy operation now that the ordering is the same.
```C++
size_t i = nBits >> 3;
vchRet.insert(vchRet.end(), m_addr.begin(), m_addr.begin() + i);
nBits &= 0x7;
```
(untested and possibly completely broken)

sipa replied:
I believe that's correct.

vasild replied:
Done - [diff together with other suggestions](https://github.com/bitcoin/bitcoin/compare/8b7f2c13b67ae84a59f29a7cffe243055a2b93d9..86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3).

sipa commented on src/netaddress.cpp:
```diff
@@ -764,53 +766,83 @@ CSubNet::CSubNet():
+    if (!valid) {
+        return;
+    }
+    // Check if `mask` contains 1-bits after 0-bits (which is an invalid netmask).
```
I don't think this is correct, for two reasons:

* We technically support non-CIDR subnets (which are allowed to contain an arbitrary mix of 0 and 1 bits).
* This test isn't even implementing whether the mask is CIDR, as it's only checking whether no 1-bits-after-0-bits occur within one byte of the mask, rather than the entire mask.

I wouldn't be opposed to dropping non-CIDR netmasks, but if we do, it should be a separate PR, I think.

vasild replied:
Uh, oh, you found a bug (the second item above)! Fixed - [diff together with other suggestions](https://github.com/bitcoin/bitcoin/compare/8b7f2c13b67ae84a59f29a7cffe243055a2b93d9..86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3).

A summary about the netmasks: https://github.com/bitcoin/bitcoin/pull/19628#issuecomment-677787025

sipa commented on src/netaddress.cpp:
```diff
@@ -822,68 +854,27 @@ bool CSubNet::Match(const CNetAddr &addr) const
-        if ((addr.ip[x] & netmask[x]) != network.ip[x])
+    assert(network.m_addr.size() == addr.m_addr.size());
+    for (size_t x = 0; x < addr.m_addr.size(); ++x)
+        if ((addr.m_addr[x] & netmask[x]) != network.m_addr[x])
```
Braces, please.

(for most rules I don't care that strongly whether you stick to the existing/surrounding style vs. follow the new style, but a braceless `if` has actually been part of the cause of a nontrivial security bug, see https://www.imperialviolet.org/2014/02/22/applebug.html)

sipa commented on src/netaddress.cpp:
```diff
@@ -822,68 +854,27 @@ bool CSubNet::Match(const CNetAddr &addr) const
     }
 
-    return network.ToString() + "/" + strNetmask;
+    return network.ToString() + strprintf("/%u", cidr);
```
Here too it seems you're dropping support for non-CIDR masks. Is that intentional?

vasild replied:
Yes, see https://github.com/bitcoin/bitcoin/pull/19628#issuecomment-677787025 - it is mentioned in the commit message and also some tests are adjusted.

vasild commented:
Addressed suggestions:

[1.](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473543192) Fixed a bug in netmask validation and extended a test to cover that case.
[2.](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473279613) Introduced `HasPrefix()` and used it where appropriate.
[3.](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473479704) Moved an assert to where `m_net` is set.
[4.](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473492495) Copy part of our address using `std::vector::insert()`.
[5.](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473546480) Added braces around single statement `for` and `if`.

vasild commented on src/netaddress.cpp:
```diff
@@ -353,13 +346,13 @@ std::string CNetAddr::ToStringIP() const
             return std::string(name);
     }
     if (IsIPv4())
```
I added { } to all places where the line containing if/for/while is modified for another reason (this one is not).

vasild commented:
A note on invalid netmasks and sub-netting non IP networks:

A netmask that contains 1-bits after 0-bits (the 1-bits are not contiguous on the left side) is invalid. IMO as invalid as "255.255.390.0" [[1]](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#Subnet_masks) [[2]](https://tools.ietf.org/html/rfc4632#section-5.1).

The code before this PR used to parse and accept such non-left-contiguous netmasks. That has been changed to not accept them because the new code is more robust that way and it is nonsensical anyway. [An IRC discussion around non-left-contiguous netmasks](http://www.erisian.com.au/bitcoin-core-dev/log-2020-04-29.html#l-241).

This PR also restricts subnetting to just IPv4 and IPv6 (subnetting of TOR or `NET_INTERNAL` is meaningless).

Those changes are mentioned in the commit message and in the PR description:

> Do not accept invalid netmasks that have 0-bits followed by 1-bits and only allow subnetting for IPv4 and IPv6

Since some comments above [[3]](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473543192) [[4]](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473547937) I looked at this carefully and managed to restore the original behavior (iuck!):

<details>
<summary>Restore invalid netmasks and non-IP subnetting on top of this PR</summary>

```diff
diff --git i/src/netaddress.cpp w/src/netaddress.cpp
index 8d73d4a58..1d4f43fcd 100644
--- i/src/netaddress.cpp
+++ w/src/netaddress.cpp
@@ -810,28 +810,16 @@ static inline int NetmaskBits(uint8_t x)
     default: return -1;
     }
 }
 
 CSubNet::CSubNet(const CNetAddr& addr, const CNetAddr& mask) : CSubNet()
 {
-    valid = (addr.IsIPv4() || addr.IsIPv6()) && addr.m_net == mask.m_net;
+    valid = addr.m_net == mask.m_net;
     if (!valid) {
         return;
     }
-    // Check if `mask` contains 1-bits after 0-bits (which is an invalid netmask).
-    bool zeros_found = false;
-    for (auto b : mask.m_addr) {
-        const int num_bits = NetmaskBits(b);
-        if (num_bits == -1 || (zeros_found && num_bits != 0)) {
-            valid = false;
-            return;
-        }
-        if (num_bits < 8) {
-            zeros_found = true;
-        }
-    }
 
     assert(mask.m_addr.size() <= sizeof(netmask));
 
     memcpy(netmask, mask.m_addr.data(), mask.m_addr.size());
 
     network = addr;
@@ -841,16 +829,13 @@ CSubNet::CSubNet(const CNetAddr& addr, const CNetAddr& mask) : CSubNet()
         network.m_addr[x] &= netmask[x];
     }
 }
 
 CSubNet::CSubNet(const CNetAddr& addr) : CSubNet()
 {
-    valid = addr.IsIPv4() || addr.IsIPv6();
-    if (!valid) {
-        return;
-    }
+    valid = true;
 
     assert(addr.m_addr.size() <= sizeof(netmask));
 
     memset(netmask, 0xFF, addr.m_addr.size());
 
     network = addr;
@@ -872,24 +857,54 @@ bool CSubNet::Match(const CNetAddr &addr) const
     }
     return true;
 }
 
 std::string CSubNet::ToString() const
 {
-    assert(network.m_addr.size() <= sizeof(netmask));
-
-    uint8_t cidr = 0;
+    /* Parse binary 1{n}0{N-n} to see if mask can be represented as /n */
+    int cidr = 0;
+    bool valid_cidr = true;
+    int size = network.m_addr.size();
+    int i;
+    for (i = 0; i < size && netmask[i] == 0xff; ++i) {
+        cidr += 8;
+    }
+    if (i < size) {
+        int bits = NetmaskBits(netmask[i]);
+        if (bits < 0) {
+            valid_cidr = false;
+        } else {
+            cidr += bits;
+        }
+        ++i;
+    }
+    for (; i < size && valid_cidr; ++i) {
+        if (netmask[i] != 0x00) {
+            valid_cidr = false;
+        }
+    }
 
-    for (size_t i = 0; i < network.m_addr.size(); ++i) {
-        if (netmask[i] == 0x00) {
-            break;
+    /* Format output */
+    std::string strNetmask;
+    if (valid_cidr) {
+        strNetmask = strprintf("%u", cidr);
+    } else {
+        if (network.IsIPv4()) {
+            strNetmask = strprintf("%u.%u.%u.%u", netmask[0], netmask[1], netmask[2], netmask[3]);
+        } else {
+            // Notice - this is not a valid representation of an IPv6 netmask.
+            // They are only defined to use CIDR notation.
+            strNetmask = strprintf("%x:%x:%x:%x:%x:%x:%x:%x",
+                             netmask[0] << 8 | netmask[1], netmask[2] << 8 | netmask[3],
+                             netmask[4] << 8 | netmask[5], netmask[6] << 8 | netmask[7],
+                             netmask[8] << 8 | netmask[9], netmask[10] << 8 | netmask[11],
+                             netmask[12] << 8 | netmask[13], netmask[14] << 8 | netmask[15]);
         }
-        cidr += NetmaskBits(netmask[i]);
     }
 
-    return network.ToString() + strprintf("/%u", cidr);
+    return network.ToString() + "/" + strNetmask;
 }
 
 bool CSubNet::IsValid() const
 {
     return valid;
 }
diff --git i/src/test/netbase_tests.cpp w/src/test/netbase_tests.cpp
index 6681c92bb..47e9f854e 100644
--- i/src/test/netbase_tests.cpp
+++ w/src/test/netbase_tests.cpp
@@ -295,17 +295,15 @@ BOOST_AUTO_TEST_CASE(subnet_test)
     subnet = ResolveSubNet("1:2:3:4:5:6:7:8/ffff:0000:0000:0000:0000:0000:0000:0000");
     BOOST_CHECK_EQUAL(subnet.ToString(), "1::/16");
     subnet = ResolveSubNet("1:2:3:4:5:6:7:8/0000:0000:0000:0000:0000:0000:0000:0000");
     BOOST_CHECK_EQUAL(subnet.ToString(), "::/0");
     // Invalid netmasks (with 1-bits after 0-bits)
     subnet = ResolveSubNet("1.2.3.4/255.255.232.0");
-    BOOST_CHECK(!subnet.IsValid());
-    subnet = ResolveSubNet("1.2.3.4/255.0.255.255");
-    BOOST_CHECK(!subnet.IsValid());
+    BOOST_CHECK_EQUAL(subnet.ToString(), "1.2.0.0/255.255.232.0");
     subnet = ResolveSubNet("1:2:3:4:5:6:7:8/ffff:ffff:ffff:fffe:ffff:ffff:ffff:ff0f");
-    BOOST_CHECK(!subnet.IsValid());
+    BOOST_CHECK_EQUAL(subnet.ToString(), "1:2:3:4:5:6:7:8/ffff:ffff:ffff:fffe:ffff:ffff:ffff:ff0f");
 }
 
 BOOST_AUTO_TEST_CASE(netbase_getgroup)
 {
     std::vector<bool> asmap; // use /16
     BOOST_CHECK(ResolveIP("127.0.0.1").GetGroup(asmap) == std::vector<unsigned char>({0})); // Local -> !Routable()
@@ -433,14 +431,13 @@ BOOST_AUTO_TEST_CASE(netbase_dont_resolve_strings_with_embedded_nul_characters)
     BOOST_CHECK(!LookupHost(std::string("127.0.0.1\0example.com\0", 22), addr, false));
     CSubNet ret;
     BOOST_CHECK(LookupSubNet(std::string("1.2.3.0/24", 10), ret));
     BOOST_CHECK(!LookupSubNet(std::string("1.2.3.0/24\0", 11), ret));
     BOOST_CHECK(!LookupSubNet(std::string("1.2.3.0/24\0example.com", 22), ret));
     BOOST_CHECK(!LookupSubNet(std::string("1.2.3.0/24\0example.com\0", 23), ret));
-    // We only do subnetting for IPv4 and IPv6
-    BOOST_CHECK(!LookupSubNet(std::string("5wyqrzbvrdsumnok.onion", 22), ret));
+    BOOST_CHECK(LookupSubNet(std::string("5wyqrzbvrdsumnok.onion", 22), ret));
     BOOST_CHECK(!LookupSubNet(std::string("5wyqrzbvrdsumnok.onion\0", 23), ret));
     BOOST_CHECK(!LookupSubNet(std::string("5wyqrzbvrdsumnok.onion\0example.com", 34), ret));
     BOOST_CHECK(!LookupSubNet(std::string("5wyqrzbvrdsumnok.onion\0example.com\0", 35), ret));
 }
 
 BOOST_AUTO_TEST_SUITE_END()
```
</details>

There are a few options:

* Leave this PR as is, or
* Amend the above patch into the sole commit of this PR and
  * Reverse-apply the patch as a separate (second) commit in this PR, or
  * Reverse-apply the patch as a separate PR after this one gets merged

What do you think?

elichai commented on src/netaddress.cpp:
```diff
@@ -3,64 +3,88 @@
 
-void CNetAddr::SetLegacyIPv6(const uint8_t ipv6[16])
+template <typename T, size_t PREFIX_LEN>
+inline bool HasPrefix(const T& obj, const uint8_t (&prefix)[PREFIX_LEN])
```
I had to google this syntax :/ 
could we please use `const std::array<uint8_t, N>&` instead? :)

elichai replied:
or just make this generic over all containers

MarcoFalke replied:
> generic over all containers

Span !?

vasild replied:
I can't figure out how to use `std::array` or `Span` and keep the usage to just `HasPrefix(m_addr, {192, 168})` (which is sweet!), so I am keeping the reference to array.

vasild replied:
> I think that can be fixed by removing the T2 template parameter, and just hardcoding it to an array of uint8_t

> I think that can be fixed by removing the `T2` template parameter, and just hardcoding it to an array of `uint8_t` (but templated size).

It does not compile either because then gcc 4.8 complained it cannot deduce the length from the initialiser list, so I [switched](https://github.com/bitcoin/bitcoin/compare/f37cde9ad6764930c0fcfefb2139e6f07d6e4353..7c34237fba619667ed1d3981d3b564ad7c25077e) to `std::array` :-/

vasild replied:
This ended up with `const std::array<uint8_t, N>&` anyway, due to GCC 4.8. Will change it back to reference to an array once the support for GCC 4.8 is removed.

ryanofsky reviewed (approved):
Code review ACK 86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3. Left some suggestions below which are minor and can be ignored. I reviewed this pretty thoroughly and didn't find any actual problems.

I can't weigh in on decision to drop support for discontiguous netmasks. Obviously that blocks current PR if it can't be done here, but otherwise this change looks very good in its current form.

vasild commented on src/netaddress.cpp:
```diff
@@ -3,64 +3,88 @@
-}
+CNetAddr::CNetAddr() {}
 
 void CNetAddr::SetIP(const CNetAddr& ipIn)
```
I agree. I considered swapping this for `operator=`, but cut it off, trying to keep the patch small. There are so many things that can be improved around this code!

jonatack reviewed (commented):
Checkpoint WIP code review and tested re-ACK f37cde9, caught up with the discussion/feedback, reviewed changes since my last review per `git diff 8b7f2c1 f37cde9` and verified a few aspects adding/removing/godbolting things, re-reviewed the code, and running a node for a few hours now.

Nit, with the latest push this code is no longer being changed but could perhaps document the `16` value in some way or other.
```diff
--- a/src/netaddress.cpp
+++ b/src/netaddress.cpp
@@ -888,12 +888,12 @@ bool CSubNet::IsValid() const
 
 bool operator==(const CSubNet& a, const CSubNet& b)
 {
-    return a.valid == b.valid && a.network == b.network && !memcmp(a.netmask, b.netmask, 16);
+    return a.valid == b.valid && a.network == b.network && !memcmp(a.netmask, b.netmask, ADDR_IPV6_SIZE);
 }
 
 bool operator<(const CSubNet& a, const CSubNet& b)
 {
-    return (a.network < b.network || (a.network == b.network && memcmp(a.netmask, b.netmask, 16) < 0));
+    return (a.network < b.network || (a.network == b.network && memcmp(a.netmask, b.netmask, ADDR_IPV6_SIZE) < 0));
 }
```

vasild commented:
Changes in the last push - fix compilation with GCC 4.8.

vasild committed:
net: don't accept non-left-contiguous netmasks

A netmask that contains 1-bits after 0-bits (the 1-bits are not
contiguous on the left side) is invalid [1] [2].

The code before this PR used to parse and accept such
non-left-contiguous netmasks. However, a coming change that will alter
`CNetAddr::ip` to have flexible size would make juggling with such
netmasks more difficult, thus drop support for those.

[1] https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#Subnet_masks
[2] https://tools.ietf.org/html/rfc4632#section-5.1

vasild committed:
net: change CNetAddr::ip to have flexible size

Before this change `CNetAddr::ip` was a fixed-size array of 16 bytes,
not being able to store larger addresses (e.g. TORv3) and encoded
smaller ones as 16-byte IPv6 addresses.

Change its type to `prevector`, so that it can hold larger addresses and
do not disguise non-IPv6 addresses as IPv6. So the IPv4 address
`1.2.3.4` is now encoded as `01020304` instead of
`00000000000000000000FFFF01020304`.

Rename `CNetAddr::ip` to `CNetAddr::m_addr` because it is not an "IP" or
"IP address" (TOR addresses are not IP addresses).

In order to preserve backward compatibility with serialization (where
e.g. `1.2.3.4` is serialized as `00000000000000000000FFFF01020304`)
introduce `CNetAddr` dedicated legacy serialize/unserialize methods.

Adjust `CSubNet` accordingly. Still use `CSubNet::netmask[]` of fixed 16
bytes, but use the first 4 for IPv4 (not the last 4). Only allow
subnetting for IPv4 and IPv6.

Co-authored-by: Carl Dong <contact@carldong.me>

vasild commented:
> Regarding the change to remove support arbitrary bitmasks as mask ... My preference would be to do it as a separate commit

Done in the last push: this PR is now split in 2 commits with the cumulative diff [equal](https://github.com/bitcoin/bitcoin/compare/7c34237fba619667ed1d3981d3b564ad7c25077e..102867c587f5f7954232fb8ed8e85cda78bb4d32) to the single commit from before (modulo the added release notes).

A further improvement would be to change `uint8_t CSubNet::netmask[16]` to `uint8_t CSubNet::m_cidr` (as a separate PR).

sipa commented on doc/release-notes.md:
```diff
@@ -135,6 +135,10 @@ Updated settings
   in future releases. Refer to the help of the affected settings `-whitebind`
   and `-whitelist` for more details. (#19191)
 
+- Netmasks that contain 1-bits after 0-bits (the 1-bits are not contiguous on
```
Perhaps point out where these are used. I think the `-rpcallowip` parameter, and the `setban` RPC are the only ones?

vasild replied:
Will do so in a followup.

sipa commented on src/netaddress.cpp:
```diff
@@ -272,13 +275,16 @@ bool CNetAddr::IsValid() const
     // so if the first length field is garbled, it reads the second batch
     // of addr misaligned by 3 bytes.
-    if (IsIPv6() && memcmp(ip, pchIPv4+3, sizeof(pchIPv4)-3) == 0)
+    if (IsIPv6() && memcmp(m_addr.data(), IPV4_IN_IPV6_PREFIX.data() + 3,
```
Not a comment on this PR (it makes sense to not change any further behavior), but 0.2.8 is over 10 years old now, and can't even communicate with current software due to the mandatory checksum change in 2012. It's probably fine to drop this code at some point.

sipa commented:
utACK 102867c587f5f7954232fb8ed8e85cda78bb4d32

ryanofsky commented on src/netaddress.cpp:
```diff
@@ -537,23 +542,29 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
-    if (nBits > 0)
-        vchRet.push_back(GetByte(15 - nStartByte) | ((1 << (8 - nBits)) - 1));
+    if (nBits > 0) {
+        assert(i < m_addr.size());
```
re: https://github.com/bitcoin/bitcoin/pull/19628#discussion_r474617351

> It is here because here we access `m_addr[i]`. I would normally use `m_addr.at(i)` if there is a chance it may get out of bounds and not write an explicit separate assert, but `prevector` does not have an `at()` method :/

Right, but if you assume the assert is false and `num_bytes >= m_addr.size()`, then the expression a few lines earlier `m_addr.begin() + num_bytes` will read from uninitialized memory. So I'm saying it might make sense to move the assert earlier there. Not very important, though.

ryanofsky reviewed (approved):
Code review ACK 102867c587f5f7954232fb8ed8e85cda78bb4d32. Just many suggested updates since last review. Thanks for following up on everything!

It's a shame this had to switch to std::array everywhere. Makes the code more verbose and fragile, but I guess it was the easiest way to work around a gcc 4.8 https://github.com/bitcoin/bitcoin/pull/19628#discussion_r474806190 compiler issue

sipa commented:
FWIW, if we just want to syntactic advantage of being able to write `HasPrefix(m_addr, {1,2,3,4})`, that could be done by making the argument `std::vector<uint8_t>`. I don't think it's worth changing that, especially as we'll probably be dropping support for GCC 4.8 soon anyway.

jonatack commented on doc/release-notes.md:
```diff
@@ -135,6 +135,10 @@ Updated settings
 
+- Netmasks that contain 1-bits after 0-bits (the 1-bits are not contiguous on
+  the left side, e.g. 255.0.255.255) are no longer accepted. They are invalid
+  according to RFC 4632.
```
Append " (#19628)"

vasild replied:
Will do so in a followup, to keep the current ACKs.

jonatack reviewed (commented):
re-ACK 102867c587f5f7954232fb8ed8e85cda78bb4d32 diff review, code review, build/tests/running bitcoind with ipv4/ipv6/onion peers

kallewoof commented on src/netaddress.cpp:
```diff
@@ -243,13 +244,15 @@ bool CNetAddr::IsTor() const { return m_net == NET_ONION; }
     // IPv6 loopback (::1/128)
     static const unsigned char pchLocal[16] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1};
-    if (IsIPv6() && memcmp(ip, pchLocal, 16) == 0)
+    if (IsIPv6() && memcmp(m_addr.data(), pchLocal, sizeof(pchLocal)) == 0) {
```
If we for a brief moment pretended that `prevector.h:258` had this line added to it:
```C++
    prevector(std::initializer_list<T> il) : prevector(il.begin(), il.end()) {}
```

then it would have been worth it to do:
```C++
static const prevector<16, uint8_t> ipv6_loopback{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1};
if (IsIPv6() && m_addr == ipv6_loopback) {
```

Alas, that prevector line would need to be added. Not sure it's worth it, or if it should be its own PR perhaps.

vasild replied:
I think it is also important to not deviate from `std::vector` so that it is not too much hassle should one decide to swap `prevector` for `std::vector` or the other way around.

Btw, `std::vector` has an initializer list constructor.

Anyway - leaving this as is.

kallewoof replied:
Yeah, it may be that my lines below (the `ipNode6`) were in fact invalid, and were the cause for the errors I saw when I did this as a prevector constructor, where I didn't initially cast 16 to `size_t`. I much prefer the constructor variant, which is also closer to `std::vector`. FTR, that was

```C++
prevector(std::initializer_list<T> il) : prevector(il.begin(), il.end()) {}
```
and the line here beomes
```C++
static prevector<ADDR_IPV6_SIZE, uint8_t> ipv6_loopback{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1};
```

vasild replied:
I wonder if we can avoid defining a local variable for IPv6 loopback address and use `IN6ADDR_LOOPBACK_INIT` or `in6addr_loopback`...

kallewoof commented on src/netaddress.cpp:
```diff
@@ -272,13 +275,16 @@ bool CNetAddr::IsValid() const
     // unspecified IPv6 address (::/128)
     unsigned char ipNone6[16] = {};
-    if (IsIPv6() && memcmp(ip, ipNone6, 16) == 0)
+    if (IsIPv6() && memcmp(m_addr.data(), ipNone6, sizeof(ipNone6)) == 0) {
```
This could be
```C++
    static prevector<16, uint8_t> ipNone6{16, 0x0};
    if (IsIPv6() && m_addr == ipNone6) {
```

vasild replied:
Right! I will apply it if I modify this PR.

`ADDR_IPV6_SIZE` is of type `size_t`, the typecast should not be needed?

kallewoof replied:
Yeah, I had originally put 16 there but realized `ADDR_IPV6_SIZE` was better.

kallewoof reviewed (commented):
Partial re-review ACK

Some nits, most of which increase the diff, so I definitely understand if you choose to ignore them.

kallewoof reviewed (commented):
ACK 102867c587f5f7954232fb8ed8e85cda78bb4d32

Reviewed changes, ran node, did IBD to ~200k, added peers etc (did not test onion routes).

vasild commented on src/netaddress.cpp:
```diff
@@ -3,79 +3,90 @@
 
-void CNetAddr::SetLegacyIPv6(const uint8_t ipv6[16])
+template <typename T1, size_t PREFIX_LEN>
+inline bool HasPrefix(const T1& obj, const std::array<uint8_t, PREFIX_LEN>& prefix)
```
This was `T` initially, then was changed to `T1` and `T2`, then I dropped the second one but forgot to rename `T1` to `T`. Will take are about it when touching this after GCC 4.8 is no longer supported.

vasild commented:
This, I think, is settled and gathered some ACKs.

Further changes to be done as a followup:

* Change `uint8_t CSubNet::netmask[16]` to `uint8_t CSubNet::m_cidr` and other improvements around `CSubNet` that are highly desirable, but not required for this PR, thus not included in it.
* Elaborate on the release notes (open this [diff](https://github.com/vasild/bitcoin/commit/b45af8c0942b873fa2ac11719b05ca6ead5378f3) as new PR).
* Use the cleaner `HasPrefix(m_addr, {1, 2, 3})` once support for GCC 4.8 is dropped. Also `s/T1/T/` in the template.

MarcoFalke commented:
Concept ACK 102867c587f5f7954232fb8ed8e85cda78bb4d32

MarcoFalke commented:
(edited OP to remove html comments)

vasild commented:
Next PR on the road to TORv3: https://github.com/bitcoin/bitcoin/pull/19845
